CMakeLists.txt

```diff
@@ -332,12 +332,62 @@ add_custom_command(                       OUTPUT  ${syscall_macros_h}
   DEPENDS ${PROJECT_SOURCE_DIR}/scripts/gen_syscall_header.py
   )
 
-# This command is a hack to support commands that are always run. Any
-# target that depends on always_rebuild will always be rebuilt.
-add_custom_command(OUTPUT always_rebuild COMMAND cmake -E echo Building for board ${BOARD})
 
-set(syscall_list_h ${CMAKE_CURRENT_BINARY_DIR}/include/generated/syscall_list.h)
-set(syscalls_json  ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls.json)
+set(syscall_list_h   ${CMAKE_CURRENT_BINARY_DIR}/include/generated/syscall_list.h)
+set(syscalls_json    ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls.json)
+set(syscalls_subdirs ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls_subdirs.txt)
+set(syscalls_trigger ${CMAKE_CURRENT_BINARY_DIR}/misc/generated/syscalls_subdirs.trigger)
+
+# When running CMake it must be ensured that all dependencies are correctly aquired.
+execute_process(
+  COMMAND
+  ${PYTHON_EXECUTABLE}
+  ${PROJECT_SOURCE_DIR}/scripts/subfolder_list.py
+  --directory        ${PROJECT_SOURCE_DIR}/include # Walk this directory
+  --out-file         ${syscalls_subdirs}           # Write this file
+  )
+file(READ ${syscalls_subdirs} PARSE_SYSCALLS_PATHS_DEPENDS)
+string(REGEX REPLACE "\n" ";" PARSE_SYSCALLS_PATHS_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
+
+if(${CMAKE_HOST_SYSTEM_NAME} STREQUAL Windows)
+  # On windows only adding/removing files or folders will be reflected in depends.
+  # Hence adding a file requires CMake to re-run to add this file to the file list.
+  set_property(DIRECTORY APPEND PROPERTY CMAKE_CONFIGURE_DEPENDS ${PARSE_SYSCALLS_PATHS_DEPENDS})
+
+  # Also On Windows each header file must be monitored as file modifications are not reflected
+  # on directory level.
+  file(GLOB_RECURSE PARSE_SYSCALLS_HEADER_DEPENDS ${PROJECT_SOURCE_DIR}/include/*.h)
+else()
+  # The syscall parsing depends on the folders in order to detect add/removed/modified files.
+  # When a folder is removed, CMake will try to find a target that creates that dependency.
+  # This command sets up the target for CMake to find.
+  add_custom_command(OUTPUT ${PARSE_SYSCALLS_PATHS_DEPENDS}
+    COMMAND ${CMAKE_COMMAND} -E echo "Preparing syscall parsing"
+    )
```

Comment:
I am not able to understand why this is needed, things appeared to work when I removed it.

Reply:
Did you try to delete a folders inside the include directory ?

Reply:
Did you try to delete a folders inside the include directory ?

The reason is that in case the build depend on a folder, let's say:
include/MyFolder, 
then the depends tells CMake that a include/MyFolder exists to depend upon.
If that folder is removed, then CMake (ninja) will try to find the target which has the output named: include/MyFolder which it cannot find, and thus it will fail.

adding a custom command specifyinf this as output gives CMake a target to execute when folder is removed.

Reply:
Thank you for the explanation.

I understand how it works now.

We need to depend on the directories, because we need to know when a folder has been created/deleted.

But if we depend on a directory and it get's deleted the build system will report that there is a missing dependency and no way to create it. So we lie to the build system and claim that the directories can be created if a certain command is run.